gnosis/safe-react
# Migrate to chains endpoint

## What it solves
Resolves #2630

Chains are now loaded from the CGW! 🙌 

`<Root>` (and `setupTests.js`) now calls the `loadChains()` method which stores the `results[]: ChainInfo[]` in a new 'local' store, `src/config/cache/chains.ts`. (Importing the Redux store directly caused circular dependencies.) The `chains` cache type mirrors what is returned from the `CGW.results`.

The Redux `config` reducer only stores the `chainId`, which is used purely to rerender the app on chain changes. The selector to access this is called `currentChainId`. (A duplicate `_chainId` is stored under a variable in `src/config/index.ts`, used as reference for finding the current chain in `chains` config 'selectors'.)

In components use the `currentChainId` selector, outside use the `_getChainId()` method (unless you have direct `store.getState()` access, i.e. in a middleware, etc.

When outside of components, it is advisable to use all 'selectors' in `src/config/index.ts` which access the `chains` cache. The name of said 'selectors' match that of the keys returned by the CGW in order to remain as coherent as possible. (Predominantly now referring to `network` as `chain`.)

Although they are generally `string` type, types have been created that inherit what is returned from the CGW  in order to remain as agnostic as possible. The following are exported:

```ts
type ShortName = ChainInfo['shortName'] // (previously `SHORT_NAME`)
type ChainId = ChainInfo['chainId'] // (previously `ETHEREUM_NETWORK`)
```

We still have a `CHAIN_ID` enum locally (previously named `ETHEREUM_NETWORK` ). **This should be used sparingly. It is, of course, only manually updated and cannot be influenced by the CGW.** It is only kept for the edge cases in which we want to specify a chain id. Instead of `'4'`, we can use `CHAIN_ID.RINKEBY`.

 **It is advisable to check all new local config 'selectors' and updated types for future reference.** Feel free to assign me as a reviewer over the coming week(s) if you touch the config. @gnosis/safe-web

## Notable changes

- `web3` is loaded with a readonly instance **after** the chains have loaded.
- Instances of `nativeCoin.address` are no longer returned from the CGW because they are always `ZERO_ADDRESS`. A new `getNativeCurrencyAddress()` method is used in place instead.
- Avalanche and Optimistic names have spaces in them. 'network root routes' now add hyphens between words. In the example of Avalanche, in order to launch the app into the network, the network route would be `/app/avalanche-mainnet`.
- When web3 instances were used for their utils, they were exchanges for eth-utils instead.

## How to test it
Upon load, whilst the preloader animation is shown, the chains are loaded from the CGW. After successful load, the networks **AVAILABLE IN THE ENVIRONMENT**are available. The app should function as normal.

**On staging, the chains that will ALWAYS be available are: Mainnet, Rinkeby and Polygon. For the sake of testing Avalanche and Optimistic, the production transaction services were added to the staging CGW.** @liliya-soroka @francovenica 

## Possible bugs
As only certain networks are available in staging and vice versa, there are a few visual inconsistencies, i.e. when the wallet is connected to a network that isn't in the network selector, the provider network will not show the text as it is not loaded from the CGW.
